Migration guide for IBMBackend.run

merav-aharoni opened the pull request:
<!--
⚠️ The pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Migration guide for transferring `IBMBackend.run()` from `qiskit_ibm_provider` to `qiskit_ibm_runtime`.


### Details and comments
Fixes #1181

drew-distefano commented:
@beckykd can you please review? This supports backend.run moving over to `qiskit-ibm-runtime` so we can sunset `qiskit-ibm-provider`. We plan to have documentation in place by the 24th, which is the research program users migrate over to IBM Cloud.

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,91 @@
+
+This guide describes how to migrate code that implemented ``backend.run()``
+using Qiskit IBM Provider (the ``qiskit_ibm_provider`` package) to use
+Qiskit IBM Runtime (``qiskit_ibm_runtime`` package).
```
```suggestion
Qiskit IBM Runtime (the ``qiskit_ibm_runtime`` package).
```

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,91 @@
+
+Migration guide: using ``backend.run()`` in ``qiskit_ibm_runtime``
+==================================================================
+
```
Could you add a short introduction here?  It should explain qiskit_ibm_provider vs qiskit_ibm_runtime and why you need to migrate.

merav-aharoni replied:
Added. Please have a look.

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,91 @@
+    circuit.cx(0, 1)
+    circuit.measure_all()
+
+In the Provider, the code is:
```
```suggestion
In Provider, the code is:
```

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,91 @@
+    job = backend.run(transpiled_circuit)
+    print(job.result())
+
+In Runtime, the code will be:
```
```suggestion
In Runtime, the code is:
```

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,91 @@
+        print(job2.session_id)
+    backend.cancel_session()
+
+The Session for ``Primitives`` (``Sampler`` and ``Estimator``) is currently different than
```
```suggestion
Using a session with primitives (``Sampler`` and ``Estimator``) is different from using a session 
```

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,91 @@
+    backend.cancel_session()
+
+The Session for ``Primitives`` (``Sampler`` and ``Estimator``) is currently different than
+the Session for ``IBMBackend``. Therefore, we cannot run a primitive and a backend
```
```suggestion
in ``IBMBackend``. Therefore, we cannot run a primitive and a backend
```

beckykd replied:
I don't understand  "...we cannot run a primitive and a backend using a single session". What does it mean to run a backend?

beckykd replied:
Maybe it should be something like "we cannot run a primitive and use backend.run() within a single session,."

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,91 @@
+
+The Session for ``Primitives`` (``Sampler`` and ``Estimator``) is currently different than
+the Session for ``IBMBackend``. Therefore, we cannot run a primitive and a backend
+using a single Session.
```
```suggestion
using a single session.
```

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,91 @@
+the Session for ``IBMBackend``. Therefore, we cannot run a primitive and a backend
+using a single Session.
+
+**Example 3: Primitive Session containing backend.run:**
```
```suggestion
**Example 3: Primitive session containing backend.run:**
```

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,89 @@
+    with Session(backend=backend) as session:
+        sampler = Sampler(session=session)
+        job1 = sampler.run(transpiled_circuit)
+        job2 = backend.run(transpiled_circuit) # runs outside the session
```
I don't understand what makes job2 run outside the session.  Is it just 'kicked out' of the session because it uses backend.run? Why would it have a session_id (from the print command) if it runs outside the session?

merav-aharoni replied:
I added an explanation and comments in the code. Please check if they are helpful.

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,89 @@
+        print(job1.session_id)
+        print(job2.session_id)
+
+**Example 4: Backend Session containing Sampler:**
```
```suggestion
**Example 4: Backend session containing Sampler:**
```

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,89 @@
+
+    with backend.open_session() as session:
+        sampler = Sampler(backend=backend)
+        job1 = sampler.run(transpiled_circuit)  # runs outside the session
```
Similarly, I don't understand why this runs outside the session.  Is it because it was instantiated with backend.open_session(), so that session just won't accept jobs that don't use backend.run?

merav-aharoni replied:
See above.

merav-aharoni commented:
@beckykd - thanks for your helpful comments. I accepted all your suggestions and added some explanations. I am not entirely sure how to write keywords, such as `Sampler`, `IBMBackend`, etc. I added `` to all of these. Please let me know what you think.

merav-aharoni commented:
@kt474 - I'd appreciate your review as well, more for the code snippets.

kt474 commented:
> @kt474 - I'd appreciate your review as well, more for the code snippets.

Snippets look good to me, I'll leave the final review for @beckykd

drew-distefano commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,98 @@
+
+In this release, we add support for running custom circuits using ``IBMBackend.run()`` in Runtime,
+so users can run all programs through Runtime.
+We eventually plan to deprecate Provider.
```
I would leave this note out, as I don't think we've officially announced this yet.

merav-aharoni replied:
@drew-distefano - Removed the last line. Please have a look.

drew-distefano reviewed (changes requested):
Seems very thorough. Please see comment for suggested change ...

beckykd commented:
> @beckykd - thanks for your helpful comments. I accepted all your suggestions and added some explanations. I am not entirely sure how to write keywords, such as `Sampler`, `IBMBackend`, etc. I added `` to all of these. Please let me know what you think.

Thanks! Keywords should be in single quotes.

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,97 @@
+    backend.cancel_session()
+
+The current implementation of Session is different for ``IBMBackend`` than that of Primitives.
+Therefore, we cannot run a primitive and a backend using a single session. This will be remediated
```
I know that it's more parallel to compare "sampler" to "backend", but I put "backend.run" because the phrase "run a backend" is just so odd :)

beckykd commented on docs/migrate/backend_run_migration_guide.rst:
```diff
@@ -0,0 +1,97 @@
+**Example 4: Backend session containing Sampler:**
+
+In this example, ``backend`` is run within a session, but ``sampler`` is run independently
+of ``session``.
```
```suggestion
of the session.
```

beckykd reviewed (approved):
I think this is good to go, after making the suggested updates.

kt474 reviewed (approved):
Addressed the remaining comments
Thanks @merav-aharoni!
